pkg/gatewayserver/io/udp/firewall_ratelimit.go

```diff
@@ -0,0 +1,93 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package udp
+
+import (
+	"sync"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	encoding "go.thethings.network/lorawan-stack/pkg/ttnpb/udp"
+)
+
+var (
+	// defaultMessages is the default number of messages to use for rate limiting.
+	defaultMessages = 10
+
+	// defaultThreshold is the duration after which a new packet will be accepted.
+	defaultThreshold = 10 * time.Millisecond
+)
+
+type rateLimitingFirewall struct {
+	f Firewall
+
+	m sync.Map // eui to incoming
+
+	messages  int
+	threshold time.Duration
+}
+
+// NewRateLimitingFirewall returns a Firewall with rate limiting capabilities.
+func NewRateLimitingFirewall(firewall Firewall, messages int, threshold time.Duration) Firewall {
+	if messages == 0 {
+		messages = defaultMessages
+	}
+	if threshold == 0 {
+		threshold = defaultThreshold
+	}
```

Comment:
These checks are useful in libraries with a public API

If there's sanity checking for bad config, it should return an error here. But I don't think that is all necessary and that you can trust the values

Reply:
Removed, GS default config should suffice